Preliminary support for 'R' mode

Reflejo opened the pull request:
Hey, this is a preliminary support for REPLACE mode. I moved the "oneChar" boolean to a new ReplaceEvaluator. It works but there are a couple of things missing:

* ~~'u' should undo the entire block not char by char.~~
* When replacing, we should never replacing the newline (end of line). This is an existing bug on the 'r' mode as well.
* When entering "backspace" on R mode, vim shows the previous char instead (again, this also is a problem with the existing 'r' mode)

Other than, I think it's a good idea to create a new evaluator for the Replace Mode instead of having a bool on insert. What do you guys think?

Keithbsmiley commented on XVim/XVimReplaceEvaluator.m:
```diff
@@ -0,0 +1,95 @@
+}
+
+- (NSString*)modeString{
+	return @"-- REPLACE --";
```
In vim just pressing `r` doesn't show a new mode string. This could just be:

```objective-c
if (_oneCharMode) {
    return @"";
} else {
    return @"-- REPLACE --";
}
```

To produce that same behavior

Reflejo replied:
Done.

Keithbsmiley commented on XVim/XVimReplaceEvaluator.m:
```diff
@@ -0,0 +1,95 @@
+	return @"-- REPLACE --";
+}
+
+- (void)abc {
```
What's the deal with this method?

Reflejo replied:
No idea how that ended up there. This should be `- (void)repeatBlockText`. Fixed.

Keithbsmiley commented on XVim/XVimReplaceEvaluator.m:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    if (nextEvaluator == self && nil == keySelector){
+        if (_oneCharMode || [self windowShouldReceive:keySelector]) {
```
Is `windowShouldReceive:` implemented somewhere?

Reflejo replied:
Yes. It's implemented in the `XVimInsertEvaluator`. Which is its parent class.

Keithbsmiley replied:
Cool, I didn't see you inherited from that.

Keithbsmiley commented on XVim/XVimReplaceEvaluator.m:
```diff
@@ -0,0 +1,95 @@
+    SEL keySelector = [keyStroke selectorForInstance:self];
+    if (keySelector){
+        nextEvaluator = [self performSelector:keySelector];
+    }else if(self.movementKeyPressed){
```
Movement keys don't seem to be accepted in either replace mode? Can this movement stuff just be removed? This would simplify these conditions too.

Reflejo replied:
Removed.

Keithbsmiley commented on XVim/XVimReplaceEvaluator.m:
```diff
@@ -0,0 +1,95 @@
+    XVimEvaluator *nextEvaluator = self;
+    SEL keySelector = [keyStroke selectorForInstance:self];
+    if (keySelector){
+        nextEvaluator = [self performSelector:keySelector];
```
Might want to wrap this in the pragma blocks if this is throwing the leaks warning.

Reflejo replied:
I don't see the warning. Maybe it's suppressed globally in the project flags?

Keithbsmiley replied:
Awesome, I just saw those pragmas elsewhere.

Keithbsmiley commented on XVim/XVimReplaceEvaluator.m:
```diff
@@ -0,0 +1,95 @@
+@end
+
+@implementation XVimReplaceEvaluator {
+    BOOL _oneCharMode;
```
Can you just make this a property instead? Seems like these not being properties is left over from supporting earlier versions of stuff.

Reflejo replied:
Done.

Keithbsmiley commented on XVim/XVimReplaceEvaluator.m:
```diff
@@ -0,0 +1,95 @@
+- (NSString*)modeString{
+	return @"-- REPLACE --";
+}
+
```
I think you want to implement `-mode` as well. Which probably means adding a value to the `XVIM_MODE` enum.

Reflejo replied:
I don't think it's necessary because it's a subclass of  `XVimInsertEvaluator`. The mode will be "INSERT" which has the right behavior.

Keithbsmiley replied:
Ok cool. I'm not sure where that mode is read from so I don't know how much that actually matters.

Keithbsmiley commented on XVim/XVimReplaceEvaluator.m:
```diff
@@ -0,0 +1,95 @@
+@property (nonatomic) BOOL movementKeyPressed;
+
+- (NSString*)insertedText;
+- (BOOL)windowShouldReceive:(SEL)keySelector;
```
Probably want to implement or remove these 2 methods.

Reflejo replied:
These two methods are implemented in the parent class:  `XVimInsertEvaluator`.

Keithbsmiley replied:
:+1:

Keithbsmiley commented:
I had a ton of little comments but this is definitely the approach I would take. I think this is much nicer than having `r` implemented in the insert evaluator.

pebble8888 commented:
I think it is a good approach.

It has some issue
- corrupt this test.
XVimMakeTestCase(text0, 5,  0, @"3rX"   , r_result2, 7, 0). please check to pass unit tests.
- inappropriate undo granularity.  
- no unit test. please write unit test case.

Reflejo commented:
> corrupt this test. XVimMakeTestCase(text0, 5, 0, @"3rX" , r_result2, 7, 0). please check to pass unit tests.

Just fixed this here: e0d55ac

> no unit test. please write unit test case.

Just added a couple of tests for R mode.

> inappropriate undo granularity.

Just fixed it. Can you check if that's the right approach? It works fine. I'm starting the undo group at `becameHandler` and ending it at `didEndHandler`

pebble8888 commented:
Hi @Reflejo
I confirmed the issues is fixed.
There are still newline and backspace issues.

Reflejo commented:
@pebble8888 Agreed, but those are existing issues with 'r' mode. I'd create two tickets for those bugs but merge this PR in the meantime.

pebble8888 commented:
I think the pull-request has enough feature
to be merged.
It has also the issue to relate multi byte characters, for example Japanese
Hiragana.
But that is a low priority issue.

On Sunday, January 4, 2015, Martín Conte Mac Donell <
notifications@github.com> wrote:

> @pebble8888 <https://github.com/pebble8888> Agreed, but those are
> existing issues with 'r' mode. I'd create two tickets for those bugs but
> merge this PR in the meantime.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/XVimProject/XVim/pull/667#issuecomment-68601959>.
>

Reflejo commented:
@JugglerShu thoughts?

JugglerShu commented:
Sorry I have been in a hospital and couldn't reply. I need to read through this PR. So plz wait till I understand everything. Thanks.

JugglerShu commented:
Sorry I should have told this first. Please make pull request to develop branch.
(I tried to merge your branch to develop but it seems there is a conflict. Let me know if it is difficult to resolve them.)

Reflejo commented:
Ok I just created a new PR here: #690
